ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -29,43 +135,80 @@ struct heap_recorder {
     .object_data = {0},
   };
 
   return recorder;
 }
 
-void heap_recorder_free(struct heap_recorder* recorder) {
-  if (recorder == NULL) {
+void heap_recorder_free(heap_recorder *heap_recorder) {
+  if (heap_recorder == NULL) {
     return;
   }
 
-  ruby_xfree(recorder);
-}
+  st_foreach(heap_recorder->object_records, st_object_record_entry_free, 0);
+  st_free_table(heap_recorder->object_records);
+
+  st_foreach(heap_recorder->heap_records, st_heap_record_entry_free, 0);
+  st_free_table(heap_recorder->heap_records);
+
+  pthread_mutex_destroy(&heap_recorder->records_mutex);
 
-// TODO: Remove when things get implemented
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wunused-parameter"
+  ruby_xfree(heap_recorder->reusable_locations);
+
+  ruby_xfree(heap_recorder);
+}
 
+// WARN: Assumes this gets called before profiler is reinitialized on the fork
 void heap_recorder_after_fork(heap_recorder *heap_recorder) {
   if (heap_recorder == NULL) {
     return;
   }
 
-  // TODO: Implement
+  // When forking, the child process gets a copy of the entire state of the parent process, minus
+  // threads.
+  //
+  // This means anything the heap recorder is tracking will still be alive after the fork and
+  // should thus be kept. Because this heap recorder implementation does not rely on free
+  // tracepoints to track liveness, any frees that happen until we fully reinitialize, will
+  // simply be noticed on next heap_recorder_flush.
+  //
+  // There is one small caveat though: fork only preserves one thread and in a Ruby app, that
+  // will be the thread holding on to the GVL. Since we support iteration on the heap recorder
+  // outside of the GVL (which implies acquiring the records_mutex lock), this means the child
+  // process may be in this weird state of having a records_mutex lock stuck in a locked
+  // state and that state having been caused by a thread that no longer exists.
+  //
+  // We can't blindly unlock records_mutex from the thread calling heap_recorder_after_fork
+  // as unlocking mutexes a thread doesn't own is undefined behaviour. What we can do is
+  // create a new lock and start using it from now on-forward. This is fine because at this
+  // point in the fork-handling logic, all tracepoints are disabled and no-one should be
+  // iterating on the recorder state so there are no writers/readers that may race with
+  // this reinitialization.
+  heap_recorder->records_mutex = (pthread_mutex_t) PTHREAD_MUTEX_INITIALIZER;
 }
 
 void start_heap_allocation_recording(heap_recorder *heap_recorder, VALUE new_obj, unsigned int weight) {
   if (heap_recorder == NULL) {
     return;
   }
 
+  VALUE ruby_obj_id = rb_obj_id(new_obj);
+  if (!FIXNUM_P(ruby_obj_id)) {
+    rb_raise(rb_eRuntimeError, "Detected a bignum object id. These are not supported by heap profiling.");
+  }
+
   heap_recorder->active_recording = (partial_heap_recording) {
-    .obj = new_obj,
+    .obj_id = FIX2LONG(ruby_obj_id),
```

Comment:
Really minor: Since we now that `ruby_obj_id` is a fixnum and thus an immediate, we could use it directly, without converting it to a long (which is just a shift anyway). That would allow us to still check for `Qnil` and whatnot in a few places.

Reply:
Ah right. It does bother me a bit from a "hygiene" perspective to store `VALUE` stuff around without GC marking them. It also makes it easy to forget/typo and start passing the actual object ref into those APIs. 

I personally don't think the ability to check for Qnil rather than 0 is a big enough pro to warrant it?